### PR TITLE
feat: Isolate iframe code from parent window

### PR DESCRIPTION
**Pull request by markdalgleish**
Each iframe is now a separate document (`frame.html`), as opposed to our previous strategy where code was executed in the parent window and then rendered into blank iframes (via [react-frame-component](https://github.com/ryanseddon/react-frame-component)).

This is actually a massive re-architecture in how code is evaluated and executed, which solves a number of issues:

- Component code is now executed within the frame, so any inspection of the browser window or browser globals will now work as expected (e.g. fixes https://github.com/seek-oss/playroom/issues/45).
- Hot reloading of CSS now works, since webpack-dev-server is now inserting styles within each frame, rather than adding them to the parent document.
- CSS no longer needs to be injected into the iframe manually on page load. The iframes now own their own styles as any page typically would.

It turned out that this model is inherently slower, so I've also included the following performance improvements that seem to make up the difference:
- The error boundary wrapping each frame no longer causes a remount whenever code changes, allowing React to properly optimise each re-render when new code is provided.
- The source JSX is now only compiled once, whereas previously we were compiling it once to validate it, then once per iframe.
- The debounce for updating code is now 500ms instead of 200ms, greatly decreasing the odds of re-rendering locking up the main thread.
- Since we now have two separate HTML files being generated, I've enabled chunking to ensure that the output is better optimised to be shared between them.

## Breaking Changes

- `FrameComponent` is longer provided with the `frameWindow` and `width` props. These are redundant now that you can just access them directly as you normally would, via `window` and `window.innerWidth`.

## Development notes

I needed to update to the latest alpha release of [html-webpack-plugin](https://github.com/jantimon/html-webpack-plugin) so that it could handle our multi-page chunking strategy correctly.

### CODE REVIEW

File: lib/makeWebpackConfig.js [hunk 1]
```diff
@@ -102,6 +105,17 @@ module.exports = async (playroomConfig, options) => {
         }
       ]
     },
+    optimization: {
+      nodeEnv: options.production ? 'production' : 'development',
+      minimize: options.production,
+      concatenateModules: options.production,
```
Comment (mattsjones) [inline lib/makeWebpackConfig.js]: The top 3 options are actually the defaults so can probably drop unless you want to be explicit.

File: src/Playroom/CatchErrors/CatchErrors.js [hunk 2]
```diff
@@ -4,33 +4,36 @@ import styles from './CatchErrors.less';
 
 export default class CatchErrors extends Component {
   static propTypes = {
+    code: PropTypes.string.isRequired,
     children: PropTypes.node.isRequired
   };
 
   state = {
     error: null,
+    erroneousCode: null,
```
Comment (mattsjones) [inline src/Playroom/CatchErrors/CatchErrors.js]: I think `invalidCode` would be easier to type lol 🤷‍♂️

File: src/Playroom/Frame.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import CatchErrors from './CatchErrors/CatchErrors';
+import RenderCode from './RenderCode/RenderCode';
+
+const themesImport = require('./themes');
+const componentsImport = require('./components');
+const frameComponentImport = require('./frameComponent');
+
+const getQueryParams = (url = window.location.href) => {
+  try {
+    const hash = url.split('#')[1] || '';
```
Comment (mattsjones) [inline src/Playroom/Frame.js]: Can we just use `window.location.hash` instead?

File: src/Playroom/Preview/Iframe.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+export default class Iframe extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      loaded: false
+    };
+  }
+
+  componentDidMount() {}
```
Comment (mattsjones) [inline src/Playroom/Preview/Iframe.js]: Remove?

File: src/Playroom/Preview/Iframe.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import React, { Component } from 'react';
+
+export default class Iframe extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
```
Comment (mattsjones) [inline src/Playroom/Preview/Iframe.js]: Needs hooks 😉

Comment (markdalgleish) [reply src/Playroom/Preview/Iframe.js]: Yep, gonna do a sweep of this at some point.

**Pull request merged**